internal/workflow/activities/unarchive.go

```diff
@@ -0,0 +1,178 @@
+package activities
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/go-logr/logr"
+	"github.com/mholt/archiver/v3"
+
+	"github.com/artefactual-sdps/enduro/internal/fsutil"
+)
+
+const UnarchiveActivityName = "unarchive-activity"
+
+// UnarchiveActivity extracts transfer files from an archive (e.g. zip, tgz).
+type UnarchiveActivity struct {
+	logger logr.Logger
+}
+
+type UnarchiveParams struct {
+	// SourcePath is the path of the transfer to be unarchived.
+	SourcePath string
+
+	// StripTopLevelDir indicates whether the top-level "container" directory of
+	// the archive should be excluded from the destination directory.
+	StripTopLevelDir bool
+}
+
+type UnarchiveResult struct {
+	// DestPath is the path to the extracted archive contents.
+	DestPath string
+
+	// IsDir is true if DestPath is a directory.
+	IsDir bool
+}
+
+func NewUnarchiveActivity(logger logr.Logger) *UnarchiveActivity {
+	return &UnarchiveActivity{logger: logger}
+}
+
+// Execute attempts to unarchive the contents of SourcePath to a temporary
+// directory. If SourcePath points to a directory or a non-archive file then the
+// path is returned, unaltered, as DestPath.
+func (a *UnarchiveActivity) Execute(ctx context.Context, params *UnarchiveParams) (*UnarchiveResult, error) {
+	a.logger.V(1).Info("Executing UnarchiveActivity", "Path", params.SourcePath)
+
+	fi, err := os.Stat(params.SourcePath)
+	if err != nil {
+		return nil, fmt.Errorf("unarchive: stat: %v", err)
+	}
+	if fi.IsDir() {
+		a.logger.V(2).Info("Unarchive: skipping directory")
+		return &UnarchiveResult{DestPath: params.SourcePath, IsDir: true}, nil
+	}
+
+	u, err := unarchiver(params.SourcePath)
+	if err != nil {
+		return nil, fmt.Errorf("unarchive: unarchiver: %v", err)
+	}
+	if u == nil {
+		// Couldn't find an unarchiver, so this is probably a regular file.
+		// Return the source path unaltered, and IsDir false.
+		a.logger.V(2).Info("Unarchive: not an archive, skipping")
+		return &UnarchiveResult{DestPath: params.SourcePath}, nil
+	}
+
+	dest := filepath.Join(filepath.Dir(params.SourcePath), "extract")
+	if err := u.Unarchive(params.SourcePath, dest); err != nil {
+		return nil, fmt.Errorf("unarchive: unarchive: %v", err)
+	}
+
+	if params.StripTopLevelDir {
+		if err = stripDirContainer(dest); err != nil {
+			return nil, fmt.Errorf("unarchive: strip top-level dir: %v", err)
+		}
+	}
+
+	if err := fsutil.SetFileModes(dest, ModeDir, ModeFile); err != nil {
+		return nil, fmt.Errorf("unarchive: %v", err)
+	}
+
+	if err := os.Remove(params.SourcePath); err != nil {
+		a.logger.V(1).Info("Unarchive: couldn't delete source archive: %v", err)
```

Comment:
should this not be an error log? if the source archive could not be deleted that might require some action because limited storage. Might make sense to alert the user here above info.

Reply:
Huh, I think at some point @sevein discouraged me from using `a.logger.Error()`, but I can't remember the details now. :thinking:

Reply:
I think that we'd want to log an error if we want to alert the system operator because there's an actionable issue that needs to be addressed. Running out of disk space is a good example but we can also assume that system monitoring should already be in place to monitor that kind of problem earlier in production. I'm ambivalent but I generally lean towards not logging errors unless strictly needed - one risk of logging too many errors is that they may ultimately be overlooked.

Reply:
I don't know if `Info()` or `Error()` is better here, but I lean slightly to `Info()` because `SourcePath()` is a temporary directory.  I think the worst case scenario is the system has no monitoring, `Remove()` fails for some reason and the "/tmp" partition fills up.  In this case a system restart will clear out any lingering transfers in "/tmp".

Reply:
> I don't know if `Info()` or `Error()` is better here, but I lean slightly to `Info()` because `SourcePath()` is a temporary directory. I think the worst case scenario is the system has no monitoring, `Remove()` fails for some reason and the "/tmp" partition fills up. In this case a system restart will clear out any lingering transfers in "/tmp".

The worst case is actually the system continues running and the tmp directories eat up all the memory and the program panics due to Out of Memory error during an activity.

Reply:
https://github.com/go-logr/logr only has `Info()` (with a verbosity level) and `Error()` methods because it's philosophy is to minimize the number of log levels. I'll switch to `Error()` here because you seem to feel strongly about it @Diogenesoftoronto.  In practice I don't expect `os.Remove()` to return an error very often, so I'm not worried about it cluttering up the error logs.

Reply:
I mean it is not just me this is from the creators of go-logr on when to use the Error():

`Info logs are things you want to tell the user which are not errors. Error logs are, well, errors. If your code receives an error from a subordinate function call and is logging that error and not returning it, use error logs.`